x-pack/plugins/discover_log_explorer/public/state_machines/log_explorer_profile/src/url_state_storage_service.ts

```diff
@@ -78,3 +105,147 @@ const extractDatasetSelectionFromIndex = ({
+        stateContainer.actions.fetchData();
+      }
+    );
+
+    // Keeps our state in sync with the url changes and makes sure it adheres to correct schema
+    const urlSubscription = stateContainer.stateStorage
+      .change$<ControlPanels>(CONTROL_PANELS_URL_KEY)
+      .subscribe((controlPanels) => {
+        if (!deepEqual(controlPanels, context.controlPanels)) {
+          send({ type: 'UPDATE_CONTROL_PANELS', controlPanels });
+        }
+      });
+
+    // Keeps the url in sync with the controls state after change
+    const inputSubscription = context.controlGroupAPI.getInput$().subscribe(({ panels }) => {
+      if (!deepEqual(panels, context.controlPanels)) {
+        send({ type: 'UPDATE_CONTROL_PANELS', controlPanels: panels });
+      }
+    });
+
+    return () => {
+      filtersSubscription.unsubscribe();
+      urlSubscription.unsubscribe();
+      inputSubscription.unsubscribe();
+    };
+  };
+
+export const updateControlPanels =
+  ({
+    stateContainer,
+  }: LogExplorerProfileUrlStateDependencies): InvokeCreator<
+    LogExplorerProfileContext,
+    LogExplorerProfileEvent
+  > =>
+  async (context, event) => {
+    if (!('controlPanels' in event)) return;
+
+    const newControlPanels = event.controlPanels || context.controlPanels;
+    const controlPanelsWithId = constructControlPanelsWithDataViewId(
+      stateContainer,
+      newControlPanels!
+    );
+
+    context.controlGroupAPI?.updateInput({ panels: controlPanelsWithId });
+
+    return controlPanelsWithId;
+  };
+
+/**
+ * Utils
+ */
+
+const constructControlPanelsWithDataViewId = (
+  stateContainer: DiscoverStateContainer,
+  newControlPanels: ControlPanels
+) => {
+  const dataView = stateContainer.internalState.getState().dataView!;
+
+  const validatedControlPanels = isValidState(newControlPanels)
+    ? newControlPanels
+    : getVisibleControlPanelsConfig(dataView);
+
+  const controlsPanelsWithId = mergeDefaultPanelsWithUrlConfig(dataView, validatedControlPanels!);
+
+  stateContainer.stateStorage.set(
+    CONTROL_PANELS_URL_KEY,
+    cleanControlPanels(controlsPanelsWithId),
+    {
+      replace: true,
+    }
+  );
+
+  return controlsPanelsWithId;
+};
+
+const isValidState = (state: ControlPanels | undefined | null): boolean => {
+  return Object.keys(state ?? {}).length > 0 && ControlPanelRT.is(state);
+};
+
+const getVisibleControlPanels = (dataView: DataView | undefined) =>
+  availableControlPanelFields.filter(
+    (panelKey) => dataView?.fields.getByName(panelKey) !== undefined
+  );
+
+export const getVisibleControlPanelsConfig = (dataView?: DataView) => {
+  return getVisibleControlPanels(dataView).reduce((panelsMap, panelKey) => {
+    const config = controlPanelConfigs[panelKey];
+
+    return { ...panelsMap, [panelKey]: config };
+  }, {} as ControlPanels);
+};
+
+const addDataViewIdToControlPanels = (controlPanels: ControlPanels, dataViewId: string = '') => {
```

Comment:
**suggestion**: what about using `_.mapValues`, `_.set` and `_.unset`from lodash to make these two utils more readable? Wdyt?
```ts
const addDataViewIdToControlPanels = (controlPanels: ControlPanels, dataViewId: string = '') => {
  return _.mapValues(controlPanels, (control) => _.set(control, 'explicitInput.id', dataViewId))
};

const cleanControlPanels = (controlPanels: ControlPanels) => {
  return _.mapValues(controlPanels, (control) => {
    _.unset(control, 'explicitInput.id');
    return control
  })
};
```